src/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets/SocketConnection.cs

```diff
@@ -0,0 +1,213 @@
+
+using Microsoft.AspNetCore.Server.Kestrel.Internal.System.Buffers;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.System.IO.Pipelines;
+using Microsoft.AspNetCore.Server.Kestrel.Transport.Abstractions;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets
+{
+    internal sealed class SocketConnection : IConnectionInformation, ITimeoutControl
+    {
+        private readonly Socket _socket;
+        private readonly SocketTransport _transport;
+        private readonly IPEndPoint _localEndPoint;
+        private readonly IPEndPoint _remoteEndPoint;
+        private IPipeWriter _input;
+        private IPipeReader _output;
+        private IList<ArraySegment<byte>> _sendBufferList;
+
+        private const int MinAllocBufferSize = 2048;        // from libuv transport
+
+        internal SocketConnection(Socket socket, SocketTransport transport)
+        {
+            Debug.Assert(socket != null);
+            Debug.Assert(transport != null);
+
+            _socket = socket;
+            _transport = transport;
+
+            _localEndPoint = (IPEndPoint)_socket.LocalEndPoint;
+            _remoteEndPoint = (IPEndPoint)_socket.RemoteEndPoint;
+        }
+
+        public async void Start(IConnectionHandler connectionHandler)
+        {
+            IConnectionContext context = connectionHandler.OnConnection(this);
+
+            _input = context.Input;
+            _output = context.Output;
+
+            // Spawn send and receive logic
+            Task receiveTask = DoReceive();
+            Task sendTask = DoSend();
+
+            // Wait for them to complete (note they won't throw exceptions)
+            await receiveTask;
+            await sendTask;
+
+            _socket.Dispose();
+        }
+        
+        private async Task DoReceive()
+        {
+            try
+            {
+                bool done = false;
+                while (!done)
+                {
+                    // Ensure we have some reasonable amount of buffer space
+                    WritableBuffer buffer = _input.Alloc(MinAllocBufferSize);
+
+                    try
+                    {
+                        int bytesReceived = await _socket.ReceiveAsync(GetArraySegment(buffer.Buffer), SocketFlags.None);
+
+                        if (bytesReceived == 0)
+                        {
+                            // EOF
+                            _input.Complete();
+                            done = true;
+                            break;
+                        }
+
+                        // record what data we filled into the buffer and push to pipe
+                        buffer.Advance(bytesReceived);
+                    }
+                    finally
+                    {
+                        var result = await buffer.FlushAsync();
```

Comment:
The interaction here is a bit more complicated that it probably should be but you need to call `context.Abort`. The fact that all of the tests pass worry me a bit because I'm sure there are some busted things.